src/availability.rs

```diff
@@ -363,10 +366,107 @@ where
                 Ok(result)
             }
             .boxed()
+        })?
+        .get("get_transaction_range_rev", move |req, state| {
+            async move {
+                let height: usize = req.integer_param("height")?;
+                let offset: usize = req.integer_param("offset")?;
+                let limit: usize = req.integer_param("limit")?;
+                let batch_size = NonZeroUsize::new(10).unwrap();
+
+                let step1 = build_range_stream_rev(height, batch_size);
+                let step2 = step1.flat_map(|range| {
+                    expand_range_to_block_range_stream::<State, Types>(range, state, timeout)
+                });
+
+                let step3 = step2.flat_map(|result| {
+                    futures::stream::iter(match result {
+                        Ok(block) => expand_block_data_to_transaction_vec(block)
+                            .into_iter()
+                            .rev()
+                            .map(Some)
+                            .collect(),
+
+                        _ => vec![None],
+                    })
+                });
+
+                let result = step3
+                    .skip(offset)
+                    .take(limit)
+                    .collect::<Vec<Option<Transaction<Types>>>>()
+                    .await;
+
+                Ok(result)
+            }
+            .boxed()
         })?;
     Ok(api)
 }
 
+// build_range_stream_rev is a helper function that builds a stream of ranges
+// to query for.  It construcs thes streams starting from the end going toward
+// 0.
+fn build_range_stream_rev(
+    end: usize,
+    chunk_size: NonZeroUsize,
+) -> impl Stream<Item = Range<usize>> {
+    let chunk_size = chunk_size.get();
+    let steps = end / chunk_size;
```

Comment:
I think we need to round up here. If `end` is 3 and `chunk_size` is 2, then `steps = 3 / 2 = 1`, and we only get one step `1..2`, missing block 0.

Reply:
On the other hand, we **could** end up with the possibility of pulling more blocks than is necessary.  But we can always cap it with a `min` with the `end` in order to avoid such a scenario... So perhaps this is fine.